language/bytecode_verifier/src/instantiation_loops.rs

```diff
@@ -0,0 +1,202 @@
+                        continue;
+                    }
+
+                    let type_actuals = &self.module.locals_signature_at(*type_actuals_idx).0;
+                    for (formal_idx, ty) in type_actuals.iter().enumerate() {
+                        let formal_idx = formal_idx as TypeParameterIndex;
+                        let node_to = self.get_or_add_node(Node(*other_func_idx, formal_idx));
+                        if let SignatureToken::TypeParameter(actual_idx) = ty {
+                            let node_from = self.get_or_add_node(Node(def.function, *actual_idx));
+                            self.graph.add_edge(node_from, node_to, Edge::Identity);
+                        } else {
+                            let type_params = Self::extract_type_parameters(ty);
+                            for type_param in type_params {
+                                let node_from =
+                                    self.get_or_add_node(Node(def.function, type_param));
+                                self.graph.add_edge(node_from, node_to, Edge::TyConApp(&ty));
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /// Computes the strongly connected components of the graph built and keep the ones that
+    /// contain at least one `TyConApp` edge. Such components indicate there exists a loop such
+    /// that an input type can get "bigger" infinitely many times along the path, also creating
+    /// infinitely many types. This is precisely the kind of constructs we want to forbid.
+    fn find_non_trivial_loops(&self) -> Vec<(Vec<NodeIndex>, Vec<EdgeIndex>)> {
+        tarjan_scc(&self.graph)
+            .into_iter()
+            .filter_map(move |nodes| {
+                let node_set: HashSet<_> = nodes.iter().cloned().collect();
+                let mut edges: Vec<EdgeIndex> = vec![];
+                let mut trivial = true;
+
+                for node_idx in &nodes {
+                    for edge in self.graph.edges(*node_idx) {
+                        if node_set.contains(&edge.target()) {
+                            edges.push(edge.id());
+                            match edge.weight() {
+                                Edge::Identity => (),
+                                Edge::TyConApp(_) => {
+                                    trivial = false;
+                                }
+                            }
+                        }
+                    }
+                }
+
+                if trivial {
+                    None
+                } else {
+                    Some((nodes, edges))
+                }
+            })
+            .collect()
+    }
+
+    pub fn verify(mut self) -> Vec<VMStatus> {
+        self.build_graph();
+        let loops = self.find_non_trivial_loops();
+
+        loops
+            .into_iter()
+            .map(|(nodes, _edges)| {
```

Comment:
Shall we include the edges, at least the interesting ones in the error message?

Reply:
That would be great.